R/impact_method2.R

```diff
@@ -0,0 +1,442 @@
+  message("Preparing temporary table <temporary_coverage_fvps>")
+  message("In progress ...")
+  
+  ## 1. touchstone specification
+  # given touchstone name, use the latest version touchstone
+  version <- DBI::dbGetQuery(con, "SELECT MAX(touchstone.version) as version FROM touchstone
+                             WHERE touchstone_name = $1", touchstone_name)
+  touchstone <- DBI::dbGetQuery(con, "SELECT id FROM touchstone
+                                WHERE touchstone_name = $1
+                                AND version = $2", list(touchstone_name, version$version))
+  
+  ## 2. creat a number table. It will be used to duplicate input data by [age_from, age_to]
+  i <- data.frame(i = 1:101, stringsAsFactors = FALSE)
+  DBI::dbWriteTable(con, "num", i, temporary = TRUE, overwrite=TRUE)
+  
+  ## 3. select minimal needed input data from db and make it age stratified - gender specific (modup is not considering gender)
+  sql <- read_sql(file_name = "coverage_pop")
+  tab <- DBI::dbGetQuery(con, sql, list(touchstone$id, 2000, 2100))
+  
+  ## 4. construct population for each vaccination activity using UN pop -
+  # this is needed to convert campaign coverage from target_pop level to UN pop level
+  # we need this to allocate campaign fvps to each targeted age
+  tab$.code <- paste(tab$vaccine, tab$activity_type, tab$country,
+                     tab$year, tab$gavi_support, tab$gender, tab$coverage,sep="><")
+  target_cohortS <- aggregate(population ~ .code, data = tab, sum, na.rm = TRUE)
+  tab <- merge_in(tab, target_cohortS, c(target_cohortS = "population"))
+  
+  ## 5. calculate age level fvps
+  tab$fvps <- NA
+  i <- is.na(tab$target)
+  # campaign
+  tab$fvps[!i] <- tab$coverage[!i] * tab$target[!i] / tab$target_cohortS[!i] * tab$population[!i]
+  # routine
+  tab$fvps[i] <- tab$coverage[i] * tab$population[i]
+  # report problems - fvps >> pop
+  tab$diff <- (tab$target - tab$target_cohortS) / tab$target_cohortS
+  write.csv(tab[tab$diff > 1., ], "problematic_campaign.csv", row.names = FALSE)
+  
+  ## 6. calculate age level coverage
+  tab$coverage <- tab$fvps / (tab$population+1) # plus one to avoid 0 pop for old age
+  tab <- tab[c("vaccine", "activity_type", "country", "year", "age", "gavi_support", "population", "coverage", "fvps")]
+  
+  ## cap fvps by population, and coverage by 1 - eg. AFG 2015 Measles campagin ; HTI Measles 2025
+  ## This operation makes sense - you cannot vaccinate more people than population
+  ## In addition, we need to cap coverage by 1 for method3 implementation (not happening for now)
+  i <- tab$coverage > 1. & !is.na(tab$coverage)
+  tab$coverage[i] <- 1
+  tab$fvps[i] <- tab$population[i]
+  
+  ## 8. make a local temporary table
+  ## so that we only manipulate pop-coverage-fvps once
+  DBI::dbWriteTable(con, "temporary_coverage_fvps", tab, temporary = TRUE, overwrite = TRUE)
+  message("Temporary table created - Success!")
+  
+}
+
+impact_output <- function(impact, method) {
```

Comment:
From the look of the comments you have, this is a time sink?  This can be made a lot more efficient I think; we can work on this as we get the reference case worked in

Reply:
Yes. It takes time to get output from a big list to a data frame. Surely, it can speed up.